src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java

```diff
@@ -0,0 +1,115 @@
+import de.tum.in.www1.artemis.repository.LearningGoalRepository;
+import de.tum.in.www1.artemis.repository.ResultRepository;
+import de.tum.in.www1.artemis.web.rest.dto.LearningGoalProgress;
+
+@Service
+public class LearningGoalService {
+
+    private final LearningGoalRepository learningGoalRepository;
+
+    private final AuthorizationCheckService authorizationCheckService;
+
+    private final ResultRepository resultRepository;
+
+    private final ParticipationService participationService;
+
+    public LearningGoalService(LearningGoalRepository learningGoalRepository, AuthorizationCheckService authorizationCheckService, ParticipationService participationService,
+            ResultRepository resultRepository) {
+        this.learningGoalRepository = learningGoalRepository;
+        this.authorizationCheckService = authorizationCheckService;
+        this.participationService = participationService;
+        this.resultRepository = resultRepository;
+    }
+
+    /**
+     * Calculate the progress in a learning goal for a specific user
+     * @param learningGoal learning goal to get the progress for
+     * @param user user to get the progress for
+     * @return progress of the user in the learning goal
+     */
+    public LearningGoalProgress getLearningGoalPerformance(LearningGoal learningGoal, User user) {
+
+        LearningGoalProgress learningGoalProgress = new LearningGoalProgress();
+        learningGoalProgress.learningGoalId = learningGoal.getId();
+        learningGoalProgress.learningGoalTitle = learningGoal.getTitle();
+        learningGoalProgress.totalPointsAchievableByStudentsInLearningGoal = 0.0;
+        learningGoalProgress.pointsAchievedByStudentInLearningGoal = 0.0;
+
+        // The progress will be calculated from a subset of the connected lecture units
+        Set<LectureUnit> filteredLectureUnits = learningGoal.getLectureUnits().parallelStream().filter(LectureUnit::isVisibleToStudents).filter(lectureUnit -> {
+            if (lectureUnit instanceof ExerciseUnit) {
+                Exercise exercise = ((ExerciseUnit) lectureUnit).getExercise();
+                return exercise.isAssessmentDueDateOver();
+            }
+            else {
+                return true;
+            }
+        }).collect(Collectors.toSet());
+
+        // In the case that two or more connected exercise units reference the same exercise, only one of them should be used in the progress calculation
+        Set<Exercise> exercisesAlreadyUsedInCalculation = new HashSet<>();
+
+        for (LectureUnit lectureUnit : filteredLectureUnits) {
+            LearningGoalProgress.LectureUnitProgress lectureUnitProgress = new LearningGoalProgress.LectureUnitProgress();
+            lectureUnitProgress.lectureUnitId = lectureUnit.getId();
+            // ToDo implement way to track progress for lecture units other than exercise units
+            if (lectureUnit instanceof ExerciseUnit) {
```

Comment:
this code is not very performant. I suggest to use the following existing queries:

```
// 1st: fetch participations, submissions and results for individual exercises
List<StudentParticipation> individualParticipations = participationService.findByStudentIdAndIndividualExercisesWithEagerSubmissionsResult(user.getId(),
        activeIndividualExercises);
// 2nd: fetch participations, submissions and results for team exercises
List<StudentParticipation> teamParticipations = participationService.findByStudentIdAndTeamExercisesWithEagerSubmissionsResult(user.getId(), activeTeamExercises);
// 3rd: merge both into one list for further processing
List<StudentParticipation> participations = Stream.concat(individualParticipations.stream(), teamParticipations.stream()).collect(Collectors.toList());
```

Reply:
Done in da86f2c21800028e9a9a960cd727cf04ede905c7